### PR TITLE
Add screenshot save feature

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

**Review by sdroege — COMMENTED** [review]
I was thinking of having a settings dialog in the app menu. It would have
1. set snapshot directory
1. set snapshot format
1. set snapshot timer length
1. set record directory (later)
1. set record format, resolution, etc (later)

The snapshot button would then only trigger a timer (cancellable, with some visual indication) and then trigger the snapshotting.

**Comment by sdroege** [comment]
@GuillaumeGomez What do you think?

**Comment by GuillaumeGomez** [comment]
Sorry, forgot to answer. And yes, it'd makes things simpler (we'll just have a filechooser dialog on the screenshot button). I'll add the setting window as soon as possible.

**Comment by sdroege** [comment]
I don't want a dialog on the snapshot button, the file chooser dialog should be in the settings :) The snapshot button should just start the timer without any further ado. It's not very user-friendly to ask every single time

**Comment by GuillaumeGomez** [comment]
Ok, so I added the settings window (I'll use a grid instead of just horizontal boxes afterwards to align a bit everything). Does it look good to you? If so I'll work on improving the look a bit.

**Comment by sdroege** [comment]
Functionality looks fine. I would make the button a simple close button though, that's what GNOME apps do :) And changes immediately have effect, there is no "saving"

File: src/main.rs [hunk 1]
```diff
@@ -29,6 +36,26 @@ macro_rules! upgrade_weak {
     };
 }
 
+#[derive(Deserialize, Serialize, Debug)]
+struct SnapshotSettings {
+    /// By default, the current one.
+    folder: PathBuf,
+    /// Either "BMP", "JPEG" or "PNG".
+    format: String,
```
Comment (sdroege) [inline src/main.rs]: Why not an enum? And no BMP please, we have no writing support for that in GStreamer :)

File: src/main.rs [hunk 2]
```diff
@@ -29,6 +36,26 @@ macro_rules! upgrade_weak {
     };
 }
 
+#[derive(Deserialize, Serialize, Debug)]
+struct SnapshotSettings {
+    /// By default, the current one.
+    folder: PathBuf,
```
Comment (sdroege) [inline src/main.rs:96]: Maybe by default this should be `glib::get_home_dir()`?

File: src/main.rs [hunk 3]
```diff
@@ -15,8 +19,11 @@ use gst::BinExt;
 use std::cell::RefCell;
 use std::env::args;
 use std::error;
+use std::path::{Path, PathBuf};
 use std::rc::{Rc, Weak};
 
+const SETTINGS_FILE: &'static str = "settings.toml";
```
Comment (sdroege) [inline src/main.rs]: Should be stored in `glib::get_user_config_dir()/application_name/settings.toml` :)

File: src/main.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/main.rs]: I'll need to check how serde works with enum but yes, good idea!

File: src/main.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/main.rs]: I discover things haha.

File: src/main.rs (hunk 1 above)
Comment (sdroege) [reply src/main.rs]: With `serde-json` it just works :)

**Comment by GuillaumeGomez** [comment]
Updated and greatly improved. Tell me what you think of it. :)

File: src/main.rs [hunk 4]
```diff
@@ -15,8 +21,12 @@ use gst::BinExt;
 use std::cell::RefCell;
 use std::env::args;
 use std::error;
+use std::fs::create_dir_all;
+use std::path::{Path, PathBuf};
```
Comment (sdroege) [inline src/main.rs]:
```
warning: unused import: `Path`                                                  
  --> src/main.rs:25:17                                                         
   |                                                                            
25 | use std::path::{Path, PathBuf};                                            
   |                 ^^^^                                                       
   |                                                                            
   = note: #[warn(unused_imports)] on by default        
```

File: src/main.rs [hunk 5]
```diff
@@ -29,6 +39,79 @@ macro_rules! upgrade_weak {
...
+    /// Timer length in seconds.
+    timer_length: usize,
+}
+
+impl Default for SnapshotSettings {
+    fn default() -> SnapshotSettings {
+        SnapshotSettings {
+            folder: glib::get_home_dir().unwrap_or_else(|| PathBuf::from(".")),
+            format: OutputFormat::default(),
+            timer_length: 3,
```
Comment (sdroege) [inline src/main.rs:108]: For some reason the timer length in the settings window is always shown as 0 for me and the +/- buttons don't work either

File: src/main.rs [hunk 6]
```diff
@@ -69,8 +152,19 @@ fn build_actions(_app: &App, application: &gtk::Application) {
...
+    let settings = gio::SimpleAction::new("settings", None);
     let about = gio::SimpleAction::new("about", None);
 
+    // When activated, show an about dialog
+    let weak_application = application.downgrade();
+    settings.connect_activate(move |_action, _parameter| {
+        let application = upgrade_weak!(weak_application);
+
+        if let Some(window) = application.get_active_window() {
+            build_screenshot_settings_window(&window);
```
Comment (sdroege) [inline src/main.rs]: screenshot is the wrong word, you mean snapshot :)

File: src/main.rs [hunk 7]
```diff
@@ -165,6 +260,196 @@ fn create_pipeline(app: &App) -> Result<(gst::Pipeline, gtk::Widget), Box<dyn er
...
+        SnapshotSettings::default()
+    };
+
+    //
+    // BUILDING UI
+    //
+    let dialog = gtk::Dialog::new_with_buttons(Some("Screenshot settings"),
+                                               Some(parent),
+                                               gtk::DialogFlags::MODAL,
+                                               &[]);
```
Comment (sdroege) [inline src/main.rs]: A close button would be useful and more consistent imho

File: src/main.rs [hunk 8]
```diff
@@ -165,6 +260,196 @@ fn create_pipeline(app: &App) -> Result<(gst::Pipeline, gtk::Widget), Box<dyn er
...
+    let grid = gtk::Grid::new();
+    grid.set_column_spacing(4);
+    grid.set_row_spacing(4);
+
+    //
+    // OUTPUT FOLDER
+    //
+    let folder_label = gtk::Label::new("Output folder");
+    let folder_hbox = gtk::Box::new(gtk::Orientation::Horizontal, 1);
+    let folder_chooser_but = gtk::Button::new_from_icon_name("folder", 1);
```
Comment (sdroege) [inline src/main.rs]: Or a `gtk::FileChooserButton`?

Comment (sdroege) [reply src/main.rs]: That also comes directly with the dialog built-in :)

File: src/main.rs [hunk 9]
```diff
@@ -165,6 +260,196 @@ fn create_pipeline(app: &App) -> Result<(gst::Pipeline, gtk::Widget), Box<dyn er
...
+        base_width: 0,
+        base_height: 0,
+        width_inc: 0,
+        height_inc: 0,
+        min_aspect: 0.,
+        max_aspect: 0.,
+        win_gravity: gdk::Gravity::West,
+    };
+    if let Some(ref w) = dialog.get_window() {
+        w.set_geometry_hints(&hints, gdk::WindowHints::MIN_SIZE | gdk::WindowHints::MAX_SIZE);
```
Comment (sdroege) [inline src/main.rs]: This all causes weird display artefacts here. Simply doing `dialog.set_resizable(false)` seems to have the intended effect without any problems

File: src/main.rs (hunk 5 above)
Comment (sdroege) [reply src/main.rs:108]: `gtk::SpinButton::new_with_range(0.0, 15.0, 1.0)` makes it work for me

File: src/main.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply src/main.rs]: Urg... That makes things so much easier!

File: src/main.rs (hunk 9 above)
Comment (GuillaumeGomez) [reply src/main.rs]: But then you can't resize the window. I had this in mind for the long paths.

Comment (sdroege) [reply src/main.rs]: The file chooser button takes care of that in a reasonable way without requiring a lot of horizontal space :) (you don't need the entry anymore either then, only the button itself)

Comment (GuillaumeGomez) [reply src/main.rs]: Oh indeed. I really wish I found this out earlier. :'(

**Comment by GuillaumeGomez** [comment]
Tadam! Once everything good on your side, I'll squash. Too much commits.

File: src/main.rs [hunk 10]
```diff
@@ -165,6 +258,144 @@ fn create_pipeline(app: &App) -> Result<(gst::Pipeline, gtk::Widget), Box<dyn er
...
+        SnapshotSettings::default()
+    };
+
+    //
+    // BUILDING UI
+    //
+    let dialog = gtk::Dialog::new_with_buttons(Some("Snapshot settings"),
+                                               Some(parent),
+                                               gtk::DialogFlags::MODAL,
+                                               &[("Close", gtk::ResponseType::Close.into())]);
```
Comment (sdroege) [inline src/main.rs:317]: The dialog could use some padding at the window borders :)

Comment (GuillaumeGomez) [reply src/main.rs:317]: Yes sir!

**Comment by sdroege** [comment]
Looks good to me, merge once there's some padding :) Thanks a lot! I'll add the actual snapshot functionality tomorrow or Sunday then

**Comment by GuillaumeGomez** [comment]
Ok, added padding. I'll merge as is. I'll now focus on adding the timer on the camera button (and have an actual implementation of the timer on it). It'll be in another PR though.

**Comment by sdroege** [comment]
> I'll now focus on adding the timer on the camera button (and have an actual implementation of the timer on it)

Simply `glib::timeout_add_seconds()` should be enough. And cancellation if pressing the button again while the timer is still running (maybe the button can be left visually downpressed for that).

What do you plan for doing the visual feedback of the timer, the countdown? Should put a seconds countdown somewhere so people know when it will trigger exactly

**Pull request merged**